### DISCUSSION TITLE
Duplicate database fails after upgrade from mysql 5.7 to 8

### DISCUSSION THREAD

**Discussion by paolomartini** [Sequel-Ace/Sequel-Ace#1907] [Q&A] [answered]
Thank you Sequel-Ace team for supporting the continuing development of Sequel Ace!

We have upgraded our DB from mySQL 5.7 to 8 (it's an AWS RDS). We are seeing no errors so far, except we just tried to duplicate a schema with Sequel Ace, including content ('duplicate database' option in the 'Database' dropdown) and we are consistently getting the following error: 'An error occurred while trying to copy the database'. The destination database gets created, but it is empty (no table, no data). This works perfectly with our mySQL 5.7 databases.

Anyone else encountering this problem with mySQL 8?

Thanks, paolo

**Comment by Jason-Morcos** [comment]
I see one sorta similar issue to this here https://github.com/Sequel-Ace/Sequel-Ace/issues/1712!

Do you see any specific error messages? Perhaps on the database log itself?

**Reply by paolomartini** [reply]
Thank you for your reply Jason!

Indeed this is similar to the ticket you reference.

Sequel Ace console gives the following:

SHOW VARIABLES LIKE 'collation_database';
SHOW DATABASES;
SHOW TABLES FROM `tmb4_dev`;
CREATE DATABASE `tmb4_dev2` DEFAULT CHARACTER SET = `utf8mb3` DEFAULT COLLATE = `utf8mb3_general_ci`;
/*!32352 SET foreign_key_checks=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!32352 SET foreign_key_checks=1 */;
/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
SHOW DATABASES;
USE `tmb4_dev2`;
SHOW VARIABLES LIKE 'character_set_database';
SHOW TABLE STATUS;
SELECT * FROM information_schema.routines WHERE routine_schema = 'tmb4_dev2' ORDER BY routine_name;

The duplicate operation fails, reliably -- I have tried numerous times, with existing and also with newly created schemas.
I certainly have enough privileges to create a new database (I am admin).
The DB logs show no errors related.

Any suggestion is most welcome!

**Reply by Jason-Morcos** [reply]
Very weird indeed - next step is probably to take a close look at the MySQL 8 changelog and look for any red flags.

Additionally, it could be helpful to run the duplicate option on a MySQL 5.7 DB and then on a MySQL 8 db and compare the Sequel Ace console outputs to see if there are any differences

**Reply by paolomartini** [reply]
Not sure what to look for in the changelog: what process does the duplicate option use? mysqldump?

This is a console dump on 5.7.42 DB (this works fine):

SHOW VARIABLES LIKE 'collation_database';
SHOW DATABASES;
SHOW TABLES FROM `newTmb3`;
CREATE DATABASE `newTmb4` DEFAULT CHARACTER SET = `utf8` DEFAULT COLLATE = `utf8_general_ci`;
/*!32352 SET foreign_key_checks=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
SHOW CREATE TABLE `newTmb3`.`access_log`;
CREATE TABLE `newTmb4`.`access_log` ......;
INSERT INTO `newTmb4`.`access_log` ......;
SHOW CREATE TABLE `newTmb3`.`api_users`;
CREATE TABLE `newTmb4`.`api_users` ......;
INSERT INTO `newTmb4`.`api_users` ......;
SHOW CREATE TABLE `newTmb3`.`batteries`;
CREATE TABLE `newTmb4`.`batteries` ......;
INSERT INTO `newTmb4`.`batteries` ......;
SHOW CREATE TABLE `newTmb3`.`results`;
CREATE TABLE `newTmb4`.`results` .......;
INSERT INTO `newTmb4`.`results` .......;
SHOW CREATE TABLE `newTmb3`.`sittings`;
CREATE TABLE `newTmb4`.`sittings` ......;
INSERT INTO `newTmb4`.`sittings` ......;
SHOW CREATE TABLE `newTmb3`.`studies`;
CREATE TABLE `newTmb4`.`studies` ......;
INSERT INTO `newTmb4`.`studies` .....;
SHOW CREATE TABLE `newTmb3`.`tests`;
CREATE TABLE `newTmb4`.`tests` ....;
INSERT INTO `newTmb4`.`tests` ......;
/*!32352 SET foreign_key_checks=1 */;
/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
SHOW DATABASES;
USE `newTmb4`;
SHOW VARIABLES LIKE 'character_set_database';
SHOW TABLE STATUS;
SELECT * FROM information_schema.routines WHERE routine_schema = 'newTmb4' ORDER BY routine_name;

This on a 8.034 DB (this fails with error 'Unable to copy database'):

SHOW VARIABLES LIKE 'collation_database';
SHOW DATABASES;
SHOW TABLES FROM `tmb4_dev`;
CREATE DATABASE `tmb4_dev2` DEFAULT CHARACTER SET = `utf8mb3` DEFAULT COLLATE = `utf8mb3_general_ci`;
/*!32352 SET foreign_key_checks=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!32352 SET foreign_key_checks=1 */;
/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
SHOW DATABASES;
USE `tmb4_dev2`;
SHOW VARIABLES LIKE 'character_set_database';
SHOW TABLE STATUS;
SELECT * FROM information_schema.routines WHERE routine_schema = 'tmb4_dev2' ORDER BY routine_name;

The 8 version seems to fail to create the tables.

**Reply by paolomartini** [reply]
BTW, Jason, we can export the database and import it into a new one just fine.

**Reply by Jason-Morcos** [reply]
Interesting, so it looks like the failure starts after we set the SQL mode - we don't run any create table statements, which jives with your notes about the DB being created empty.

Honestly not sure what's being used under the hood - I bet it is mysqldump, and I bet the output of the mysql dump looks different from 5.7 to 8. What's strange is that - that you're able to export and import the export manually

@Sequel-Ace/all I don't have a MySQL 8 setup. Anyone have a MySQL 8 to poke further at this with?

**Reply by paolomartini** [reply]
Jason,

I looked at the code and it's not mysqldump, as far as I can tell, just a series of show-create-insert statements iterated over the list of tables.

One thing that is different between import and copyTable is that import seems to issue this statement before creating the table:

SET NAMES utf8mb4

The upgrade has changed the default character set and collation, I wonder if this is the culprit.

Here are some of the relevant server variables:

character_set_client = utf8mb4
character_set_connection = utf8mb4
character_set_database = utf8mb3
character_set_filesystem = binary
character_set_results = utf8mb4
character_set_server = utf8mb4
character_set_system = utf8mb3
collation_connection = utf8mb4_0900_ai_ci
collation_database = utf8mb3_general_ci
collation_server = utf8mb4_0900_ai_ci
default_collation_for_utf8mb4 = utf8mb4_0900_ai_ci

**Reply by Jason-Morcos** [reply]
That definitely makes sense as a theory. Do you have a comparison of the server variables form 5.7 versus 8? It could be something about utf8mb3, or about the special utf8mb4_0900_ai_ci collation perhaps

**Reply by Jason-Morcos** [reply]
Additionally, I wonder if changing any of these server variables in 8 makes the export work? That could help us isolate which exactly is problematic

**Reply by robinkunde** [reply]
What's the exact engine (Aurora Mysql / MySQL / MariaDB) and engine version you're using? I can spin up an RDS instance and try to replicate the issue. Unless you could spin up one for us to test against?

**Reply by paolomartini** [reply]
Hi Robin,

It's an RDS MySQL community version 8.034 (not Aurora).
I can try spinning up a new one later tonight, but if you have bandwidth it
would be good if you could kindly try yourself, so we would know if it is
something specific to us, you never know with AWS!

Thanks so much for pitching in, we appreciate it!

Paolo

**Reply by paolomartini** [reply]
Hi Jason,

Here are the server variables for the 5.7.42 DB:

character_set_client = utf8mb4
character_set_connection = utf8mb4
character_set_database = utf8
character_set_filesystem = binary
character_set_results = utf8mb4
character_set_server = latin1
character_set_system = utf8
collation_connection = utf8mb4_general_ci
collation_database = utf8_general_ci
collation_server = latin1_swedish_ci

**Reply by paolomartini** [reply]
As a bit of context, mySQL 5.7 was EOL this past October 31st and will no longer receive security patches. AWS won't make it available for install after end February 2024 and will only offer an extended support option for pay after that date.

**Reply by robinkunde** [reply, marked as answer]
https://github.com/Sequel-Ace/Sequel-Ace/pull/1919

**Reply by paolomartini** [reply]
That's fantastic! Thank you Robin!
